### PR TITLE
test: introduces custom chai assertions

### LINKED ISSUE #184: Consider custom chai assertions
We seem to have a lot of repetitive assertion logic in tests. It may be beneficial to extend `chai`'s assertion to use common custom assertions more freely:

```js
assert.hasValidator(result.headers, 'JsonSchema')
assert.hasRealType(result.headers, 'text/plain')
```

### PR DESCRIPTION
**Pull request by artem-zakharchenko**
These changes introduce custom Chai assertions to use in Gavel unit and integration tests.

## Motivation

- To have less verbose test suites
- To have common assertions integrated into `chai` chainable operators. Common assertions include:
  - result/field validity
  - expected/real type
  - errors
- To get rid of implementation details in test, and centralize validation result references (to rely on validation result structure in a single place–custom assertions declaration)
- To enhance test exceptions with more information about the scenario (include stringified validation result and optional hints)

## GitHub

- Closes #184

## Roadmap

- [x] Implement custom assertions
  - [x] result and field validity
  - [x] validator
  - [x] expectedType
  - [x] realType
  - [x] errors / noErrors
- [x] Use `expect` instead of `assert` for better readability of custom assertions
- [x] Ensured consistency of assertions between test suites
- [x] ~~Consider using `to.not.have.validator` as a replacement for `.to.have.validator(null)`~~

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -14,6 +14,7 @@ module.exports = {
 
     // Temporary overrides. Logic to be rewritten.
     // TODO https://github.com/apiaryio/gavel.js/issues/150
-    'no-param-reassign': 'off'
+    'no-param-reassign': 'off',
+    'no-unused-expressions': 'off'
```
Comment (artem-zakharchenko) [inline .eslintrc.js]:
```js
expect(obj).to.be.valid
```
is considered a violation of this rule. Therefore, disabling the rule. Such violations are not present anywhere else in the code.

**Comment by artem-zakharchenko** [comment]
Experienced #139, restarting the build.

File: .eslintrc.js (hunk 1 above)
Comment (honzajavorek) [reply .eslintrc.js]: I think this should be then a separate "section" if the ignores:

```js
... 

// Disabled to allow expect() expressions in tests
'no-unused-expressions': 'off',

// Temporary overrides. Logic to be rewritten
'no-param-reassign': 'off',
...
```

File: test/integration/validate.test.js [hunk 2]
```diff
@@ -38,39 +18,44 @@ describe('validate', () => {
...
-      expectedType(result.fields.method, 'text/vnd.apiary.method');
-      realType(result.fields.method, 'text/vnd.apiary.method');
+      expect(result.fields.method).to.be.valid;
+      expect(result.fields.method).to.have.validator(null);
+      expect(result.fields.method).to.have.expectedType(
+        'text/vnd.apiary.method'
+      );
+      expect(result.fields.method).to.have.realType('text/vnd.apiary.method');
+
       noErrors(result.fields.method);
```
Comment (honzajavorek) [inline test/integration/validate.test.js]: `noErrors` is intentionally going to be a next step or was it forgotten?

File: test/unit/units/validateBody.test.js [hunk 3]
```diff
@@ -395,12 +435,43 @@ describe('validateBody', () => {
...
+                `At '/firstName' Missing required property: firstName`
               );
             });
           });
         });
       });
     });
   });
+
+  // describe.only('when validating Json Schema V3 (Amanda)', () => {
```
Comment (honzajavorek) [inline test/unit/units/validateBody.test.js]: dramatic chipmunk

File: test/integration/validate.test.js (hunk 2 above)
Comment (artem-zakharchenko) [reply test/integration/validate.test.js]: The next step. I haven't yet decided what would be a comfortable interface to assert Gavel's validation errors, so I left errors assertions behind.

File: test/unit/units/validateBody.test.js (hunk 3 above)
Comment (artem-zakharchenko) [reply test/unit/units/validateBody.test.js]: Removed during rebase of the existing commit.

File: .eslintrc.js (hunk 1 above)
Comment (artem-zakharchenko) [reply .eslintrc.js]: Grouped in cbfb18c2dbb9856f00e7458dcd6f940d332168e0.

**Comment by artem-zakharchenko** [comment]
I still would like to provide errors assertions as a part of this pull request, just to make it clear. Focusing on them now.

**Comment by honzajavorek** [comment]
I think it is completely okay to merge this now and add them in a separate PR, but up to you.

**Comment by artem-zakharchenko** [comment]
Custom assertions for the errors were also added, so the pull request should be complete!

I ditched an idea to have `to.not.have.validator`, because I think that conceptually means there is no `validator` property, which is not the expected behavior.

**Pull request merged**